src/shapes/ray.rs

```diff
@@ -44,56 +45,54 @@ pub struct Ray {
 }
 
 /// A struct which is returned by the `intersects_triangle` method.
+#[derive(Debug, Clone, Copy)]
 pub struct Intersection {
     /// Distance from the ray origin to the intersection point.
-    pub distance: f32,
+    pub distance: Real,
 
     /// U coordinate of the intersection.
-    pub u: f32,
+    pub u: Real,
 
     /// V coordinate of the intersection.
-    pub v: f32,
+    pub v: Real,
+
+    /// Normal of the intersection
+    pub norm: Vector3,
+
+    /// Whether the intersect was a backface
+    pub back_face: bool,
 }
 
 impl Intersection {
     /// Constructs an `Intersection`. `distance` should be set to positive infinity,
     /// if the intersection does not occur.
-    pub fn new(distance: f32, u: f32, v: f32) -> Intersection {
-        Intersection { distance, u, v }
+    pub fn new(distance: Real, u: Real, v: Real, norm: Vector3, back_face: bool) -> Intersection {
+        Intersection {
+            distance,
+            u,
+            v,
+            norm,
+            back_face,
+        }
     }
 }
 
-impl Ray {
-    /// Creates a new [`Ray`] from an `origin` and a `direction`.
-    /// `direction` will be normalized.
-    ///
-    /// # Examples
-    /// ```
-    /// use bvh::ray::Ray;
-    /// use bvh::{Point3,Vector3};
-    ///
-    /// let origin = Point3::new(0.0,0.0,0.0);
-    /// let direction = Vector3::new(1.0,0.0,0.0);
-    /// let ray = Ray::new(origin, direction);
-    ///
-    /// assert_eq!(ray.origin, origin);
-    /// assert_eq!(ray.direction, direction);
-    /// ```
-    ///
-    /// [`Ray`]: struct.Ray.html
-    ///
-    pub fn new(origin: Point3, direction: Vector3) -> Ray {
-        let direction = direction.normalize();
-        Ray {
-            origin,
-            direction,
-            inv_direction: Vector3::new(1.0 / direction.x, 1.0 / direction.y, 1.0 / direction.z),
-            sign_x: (direction.x < 0.0) as usize,
-            sign_y: (direction.y < 0.0) as usize,
-            sign_z: (direction.z < 0.0) as usize,
-        }
+/// This trait can be implemented on anything that can intersect with a `Ray`
+pub trait IntersectionRay {
```

Comment:
Do you have your own IntersectionRay impls outside of this library? I'm wondering what the performance tradeoff of using a vtable here vs an enum would be.

Also, is there a reason why IntersectionRay is only implemented for sphere+triangle, but not aabb/capsule/obb?